File: backend/internal/models/analytics.go
package models

type MedianPriceResult struct {
	Region      string `json:"region"`
	MedianPrice int64  `json:"median_price"`
}

type PriceTrendResult struct {
	Period           string `json:"period"`
	AvgPrice         int64  `json:"avg_price"`
	MedianPrice      int64  `json:"median_price"`
	TransactionCount int64  `json:"transaction_count"`
}

type AffordabilityResult struct {
	PropertyType         string  `json:"property_type"`
	AvgPrice             int64   `json:"avg_price"`
	RelativeAffordability float64 `json:"relative_affordability"`
}

type GrowthHotspotResult struct {
	Region        string  `json:"region"`
	GrowthRate    float64 `json:"growth_rate"`
	PrevMedian    int64   `json:"prev_median"`
	CurrentMedian int64   `json:"current_median"`
}

type NewBuildPremiumResult struct {
	Region         string  `json:"region"`
	NewAvg         int64   `json:"new_avg"`
	OldAvg         int64   `json:"old_avg"`
	PremiumPercent float64 `json:"premium_percent"`
}

type PropertyTypeDistributionResult struct {
	PropertyType string  `json:"property_type"`
	Count        int64   `json:"count"`
	Percentage   float64 `json:"percentage"`
}

type PriceBracketResult struct {
	Bracket    string  `json:"bracket"`
	Count      int64   `json:"count"`
	Percentage float64 `json:"percentage"`
}

type TopActiveAreaResult struct {
	Region           string `json:"region"`
	TransactionCount int64  `json:"transaction_count"`
	TotalValue       int64  `json:"total_value"`
}

type TimeRangeResult struct {
	MinYear int `json:"min_year"`
	MaxYear int `json:"max_year"`
}

File: frontend/src/query/analytics.ts
import { createQueryWrapper } from "@/utils/query";
import { defaultQueryFn } from "@/utils/query/defaultQuery";

export interface MedianPriceResult {
  region: string;
  median_price: number;
}

export interface PriceTrendResult {
  period: string;
  avg_price: number;
  median_price: number;
  transaction_count: number;
}

export interface AffordabilityResult {
  property_type: string;
  avg_price: number;
  relative_affordability: number;
}

export interface GrowthHotspotResult {
  region: string;
  growth_rate: number;
  prev_median: number;
  current_median: number;
}

export interface NewBuildPremiumResult {
  region: string;
  new_avg: number;
  old_avg: number;
  premium_percent: number;
}

export interface PropertyTypeDistributionResult {
  property_type: string;
  count: number;
  percentage: number;
}

export interface PriceBracketResult {
  bracket: string;
  count: number;
  percentage: number;
}

export interface TopActiveAreaResult {
  region: string;
  transaction_count: number;
  total_value: number;
}

export const medianPriceQuery = createQueryWrapper<
  MedianPriceResult[],
  { by: string },
  { param: { by: string } }
>({
  queryKey: (params) => ["analytics", "median-price", params.by],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/median-price?by=$by",
    })<MedianPriceResult[]>(),
  },
});

export const priceTrendQuery = createQueryWrapper<
  PriceTrendResult[],
  { interval: string },
  { param: { interval: string } }
>({
  queryKey: (params) => ["analytics", "price-trend", params.interval],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/price-trend?interval=$interval",
    })<PriceTrendResult[]>(),
  },
});

export const affordabilityQuery = createQueryWrapper<AffordabilityResult[]>({
  queryKey: () => ["analytics", "affordability"],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/affordability",
    })<AffordabilityResult[]>(),
  },
});

export const growthHotspotsQuery = createQueryWrapper<
  GrowthHotspotResult[],
  { limit: number },
  { param: { limit: number } }
>({
  queryKey: (params) => ["analytics", "growth-hotspots", params.limit],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/growth-hotspots?limit=$limit",
    })<GrowthHotspotResult[]>(),
  },
});

export const newBuildPremiumQuery = createQueryWrapper<
  NewBuildPremiumResult[],
  { by: string },
  { param: { by: string } }
>({
  queryKey: (params) => ["analytics", "new-build-premium", params.by],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/new-build-premium?by=$by",
    })<NewBuildPremiumResult[]>(),
  },
});

export const propertyTypeDistributionQuery = createQueryWrapper<
  PropertyTypeDistributionResult[]
>({
  queryKey: () => ["analytics", "property-type-distribution"],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/property-type-distribution",
    })<PropertyTypeDistributionResult[]>(),
  },
});

export const priceBracketDistributionQuery = createQueryWrapper<
  PriceBracketResult[]
>({
  queryKey: () => ["analytics", "price-bracket-distribution"],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/price-bracket-distribution",
    })<PriceBracketResult[]>(),
  },
});

export const topActiveAreasQuery = createQueryWrapper<
  TopActiveAreaResult[],
  { by: string; limit: number },
  { param: { by: string; limit: number } }
>({
  queryKey: (params) => ["analytics", "top-active-areas", params.by, params.limit],
  options: {
    queryFn: defaultQueryFn({
      url: "/analytics/top-active-areas?by=$by&limit=$limit",
    })<TopActiveAreaResult[]>(),
  },
});
